### PR TITLE
dev-amdgpu, gpu-compute, mem-ruby: Add support for writeback L2 in GPU

### PR DESCRIPTION
**Pull request by v-ramadas**
Previously, GPU L2 caches could be configured in either writeback or writethrough mode when used in an APU. However, in a CPU+dGPU system, only writethrough worked. This is mainly because in CPU+dGPU system, the CPU sends either PCI or SDMA requests to transfer data from the GPU memory to CPU. When L2 cache is configured to be writeback, the dirty data resides in L2 when CPU transfers data from GPU memory. This leads to the wrong version being transferred. A similar issue also crops up when the GPU command processor reads kernel information before kernel dispatch, only to incorrect data. This PR contains a set of commits that fix both these issues.

### CODE REVIEW

File: src/mem/ruby/protocol/GPU_VIPER-TCC.sm
Comment (mattsinc) [inline src/mem/ruby/protocol/GPU_VIPER-TCC.sm:835]: this is not my favorite format, but I guess in this case checking isDirty or something like that is not sufficient?  If so, then this is fine.

File: src/mem/ruby/protocol/MOESI_AMD_Base-dir.sm
Comment (mattsinc) [inline src/mem/ruby/protocol/MOESI_AMD_Base-dir.sm]: tabs/spacing off on this line?

File: src/mem/ruby/protocol/MOESI_AMD_Base-dir.sm
Comment (mattsinc) [inline src/mem/ruby/protocol/MOESI_AMD_Base-dir.sm]: remove excess lines

**Review by mattsinc — DISMISSED** [review]
A few minor changes but from my end this mostly looks like what we discussed.

@abmerop should look at this before we merge though

File: src/mem/ruby/protocol/MOESI_AMD_Base-dir.sm
Comment (v-ramadas) [reply src/mem/ruby/protocol/MOESI_AMD_Base-dir.sm]: Done

File: src/mem/ruby/protocol/MOESI_AMD_Base-dir.sm
Comment (v-ramadas) [reply src/mem/ruby/protocol/MOESI_AMD_Base-dir.sm]: I/Nagendra added an additional if() condition to check if L2 is WB. The additional spaces are from that, I think.

File: src/mem/ruby/protocol/GPU_VIPER-TCC.sm
Comment (v-ramadas) [reply src/mem/ruby/protocol/GPU_VIPER-TCC.sm:835]: The TBE/cache entry's dirty flag doesn't seem to be set anywhere. I would have to set that to true/false in transitions before I can use it here. Would you prefer I do that, or is this sufficient?

File: src/gpu-compute/compute_unit.cc
Comment (abmerop) [inline src/gpu-compute/compute_unit.cc]: This is fine, however it might be cleaner to add something to the SQCPort::SenderState instead indicating it came from the command processor. I am not sure the origins of this flag, but my assumption has been it indicates that address translation should be bypassed. The current way of reading the kernel code structure uses this flag, so I can see why you did this.

File: src/mem/ruby/system/RubySystem.cc
Comment (abmerop) [inline src/mem/ruby/system/RubySystem.cc]: This can be dropped, it looks like no there are changes to this file.

File: src/mem/ruby/protocol/MOESI_AMD_Base-dir.sm
Comment (abmerop) [inline src/mem/ruby/protocol/MOESI_AMD_Base-dir.sm:1667]: This seems suspect. Is there some case where duplicate memory requests get sent and we ignore their responses?

File: src/mem/ruby/protocol/GPU_VIPER-TCC.sm
Comment (abmerop) [inline src/mem/ruby/protocol/GPU_VIPER-TCC.sm]: Could you explain this part? Do PrbInv requests only happen in write-through cache? I see elsewhere that `L2isWB` is used instead to check this, so does that imply PrbInv can happen in write-back cache too and the comment needs to be updated?

File: src/mem/packet.hh
Comment (abmerop) [inline src/mem/packet.hh]: Historically these have gone in the request class.  Does it need to be part of Packet class (i.e., can you have a Packet with a nullptr Request?)

**Review by abmerop — COMMENTED** [review]
I really like adding the timing read for the kernel code object!  I believe this is the last place the GPU model is using a functional access in a place where it is not intentional (e.g., a performance optimization).

I think I mostly follow the protocol changes but not completely so I could use some clarification there.

File: src/mem/packet.hh
Comment (powerjg) [reply src/mem/packet.hh]: To add on here, I would *strongly* prefer not changing Packet *or* Request unless absolutely necessary. There should be a very high bar for changing these widely used objects. Why do we need a special functional access flag for the GPU?

Comment (abmerop) [reply src/mem/packet.hh]: I can try to answer based on the comment in RubyPort.cc... This is related to the multiple Networks in one RubySystem which uses the requestor ID to know if requests should go to the GPU or CPU network. It defaults to 0 which is an unmatched network = CPU side. To know the network at the RubyPort, some misc. object accessible by the RubyPort must know which network to pick. Packet/Request have been the best interface so far.

This is mainly because of the way *I* implemented that multiple network support. The multiple-RubySystem supersedes this but would require moving to that...  In the long term everything related to the multiple Networks will need a refactor, I think.  I am aware of some folks using it, but I don't know if they can use multiple-RubySystem. Right now this is the common way to deal with multiple networks in Ruby.

Comment (powerjg) [reply src/mem/packet.hh]: I suppose my question is whether this is duplicate information. E.g., can we infer that it's a GPU functional access from (1) it's functional, and (2) it is in the 1 network?

File: src/mem/ruby/protocol/GPU_VIPER-TCC.sm
Comment (v-ramadas) [reply src/mem/ruby/protocol/GPU_VIPER-TCC.sm]: I've updated the comments to remove the write-through/writeback part. It is more generic now. Thanks for pointing this out!

File: src/mem/ruby/system/RubySystem.cc
Comment (v-ramadas) [reply src/mem/ruby/system/RubySystem.cc]: Removed the commented out section. I'm unable to revert the change to the DPRINTF. The gem5 style checker complains about the old DPRINTF() having an invalid line length and wants me to put the line break in. Is there a way to get past this and revert it?

File: src/mem/ruby/protocol/MOESI_AMD_Base-dir.sm
Comment (v-ramadas) [reply src/mem/ruby/protocol/MOESI_AMD_Base-dir.sm:1667]: The protocol changes are intended to work as follows when the cache line receives a DMAReadWB event trigger (only triggered if L2 is WB):

- The cache line moves from U to BL2. It checks L3 cache (if size is non-zero), probes L2 caches, and sends out a request to memory simultaneously.
- If it is a hit at L3 (line 1625) or if L2 sends back data (line 1630), the line transitions to BL2_Pm (got memory but waiting for probes to complete). Any data from memory is now ignored (line 1666)
- If it is not a hit at L3 and if L2 sends probe ack without data (line 1610), the cache waits for data from memory. If data from memory arrives before ProbeAcksComplete, the transition at line 1660 is called first and the transition at line 1650 is called when all probes are received
- If it is not a hit at L3 and if none of the L2s send data but send ProbeAcksComplete before memory request returns, the transition at line 1643 is called and the line moves to BL2_M state (waiting for mem). Then, when data from memory arrives, the transition at line 1669 is called.

There are four possible states for the cacheline when mem data returns: if it is in BL2 or BL2_M, the data is written to TBE and any dirty data in L2/L3 overwrites the TBE later. If it already received dirty data (in BL2_Pm) the mem data is ignored. If the line is in U (already received dirty data and all acks) the mem data is ignored.

I am unable to think of any other case. Please let me know if you can think of any, and I will add that too!

File: src/mem/ruby/system/RubySystem.cc
Comment (abmerop) [reply src/mem/ruby/system/RubySystem.cc]: There *is* however I think it's fine to keep this.  Someday it will be caught by clang-format anyway :).

File: src/gpu-compute/compute_unit.cc
Comment (v-ramadas) [reply src/gpu-compute/compute_unit.cc]: That sounds like a good idea! I'm currently working on it and will update the commit soon

Comment (v-ramadas) [reply src/gpu-compute/compute_unit.cc]: I've pushed this change. I added a isKernDispatch field to SQCPort::SenderState that tracks whether the request is a kernel dispatch or not

File: src/mem/packet.hh
Comment (v-ramadas) [reply src/mem/packet.hh]: The issue I faced when making this fix was that the GPU functional accesses go through network 0 (CPU network) by default. I use this flag to make packets use one of the GPU requestor IDs so that it is injected into network 1 when it reaches RubyPort instead of network 0. I couldn't find any other structure that is shared between Ruby and the GPU compute units that would help me with this.

File: src/mem/ruby/protocol/GPU_VIPER-TCC.sm
Comment (mattsinc) [reply src/mem/ruby/protocol/GPU_VIPER-TCC.sm:835]: let it be then

File: src/mem/ruby/protocol/MOESI_AMD_Base-dir.sm
Comment (mattsinc) [reply src/mem/ruby/protocol/MOESI_AMD_Base-dir.sm]: I'm not sure how that can be for line 311?  But if it's an optical illusion, ok

File: src/mem/ruby/system/RubySystem.cc
Comment (mattsinc) [reply src/mem/ruby/system/RubySystem.cc]: Going to mark this resolved

File: src/mem/ruby/protocol/MOESI_AMD_Base-dir.sm
Comment (v-ramadas) [reply src/mem/ruby/protocol/MOESI_AMD_Base-dir.sm]: Surprisingly, I still see the right indentation in my github, vim views. But when I open the file in vim, I see that it uses a tab instead of spaces. Maybe a different tab/space configurations cause it to be formatted differently in different environments? I have now changed the tab to space. Hope that fixes it!

File: src/mem/packet.hh
Comment (v-ramadas) [reply src/mem/packet.hh]: > Historically these have gone in the request class. Does it need to be part of Packet class (i.e., can you have a Packet with a nullptr Request?)

Moved it to the request class.

Comment (mattsinc) [reply src/mem/packet.hh]: My sense is that @v-ramadas is saying the answer to your question is no, @powerjg .  Or at least absent the complete working @abmerop is proposing, the answer is no.  So unless you all have an alternative, I think this is the best option we could think of.  Let us know ...

File: src/mem/ruby/protocol/MOESI_AMD_Base-dir.sm
Comment (mattsinc) [reply src/mem/ruby/protocol/MOESI_AMD_Base-dir.sm:1667]: @abmerop just checking if this addresses your concern?

**Review by mattsinc — APPROVED** [review]
Approving so testing can happen in parallel with @abmerop and @powerjg re-reviewing

File: src/mem/ruby/protocol/MOESI_AMD_Base-dir.sm
Comment (abmerop) [reply src/mem/ruby/protocol/MOESI_AMD_Base-dir.sm:1667]: Thanks for the explanation!  It should then be impossible to get MemData in U state, correct?  This is the transition that concerns me.  MemData in BL2_Pm state makes sense.  If that is true, then the change would be `{BL2_Pm, U}` -> `BL2_Pm`.

The concern is it might be hiding some other bugs (potentially entirely unrelated to this PR) which go to U state when it should be waiting for memory.

**Review by powerjg — APPROVED** [review]
I still have reservations about modifying the request for this one small use case. My gut says there's a better way to do this. However, I don't understand the problem well enough to suggest anything specific; so, this change is OK with me.

**Comment by abmerop** [comment]
> I still have reservations about modifying the request for this one small use case. My gut says there's a better way to do this. However, I don't understand the problem well enough to suggest anything specific; so, this change is OK with me.

If it makes it better, this won't be needed anymore after the VIPER board.  That is using multiple RubySystems instead of multiple RubyNetworks (aka disjoint network).  In the multiple RubySystems case, there is no ambiguity which network functional reads go to.

**Comment by powerjg** [comment]
So should we not merge this?

**Comment by abmerop** [comment]
We should merge this. The old configs are not yet deprecated and still using multiple RubyNetworks.

File: src/mem/ruby/protocol/MOESI_AMD_Base-dir.sm
Comment (v-ramadas) [reply src/mem/ruby/protocol/MOESI_AMD_Base-dir.sm:1667]: It would be possible to get MemData in U state if the memory request returns after the cache probes complete. A slow memory request, in this case, would then come to a cache state in U. I now wonder if this would introduce a scenario where a new request comes in before MemData returns, and lead to an inconsistency. I will update the protocol to handle this scenario too.

Comment (abmerop) [reply src/mem/ruby/protocol/MOESI_AMD_Base-dir.sm:1667]: Sorry to belabor this, but I am still missing something.  In the case you describe wouldn't be in `BL2_M`? I.e., the state which is "got probes, waiting for memory"?

Comment (v-ramadas) [reply src/mem/ruby/protocol/MOESI_AMD_Base-dir.sm:1667]: The line would be in BL2_M only if none of the probes return dirty data. If a probe did, then the line moves to BL2_Pm.

There are two paths the sequence of transitions could take. When the state is in BL2, it is waiting for both probes from caches and response from memory. If the cache probes return dirty data, the line transitions to BL2_Pm. From here, either all the cache probes could return first (BL2_Pm->U) or memory could return first (remains in BL2_PM and moves to U once other probes return). If none of the cache probes returns dirty data, the line could move to BL2_M once all probes return but MemData hasn't or move to BL2_Pm if MemData comes first (BL2_Pm->U once probes return).

I will add more comments, change the state descriptions, and use a different name for the CPUPrbRespWB event (change it to CPUPrbRespDirty)

**Pull request merged**
